### PR TITLE
Fix type error

### PR DESCRIPTION
**Pull request by timlyeee**
According to ts test, list of Scheduler may store Node object and Component object, as if it contains id, uuid and update function. For c++ this logic is unacceptable.
To adapt this need(Temporatory), add SchedulerPointer to encapsulize the objects.
* Add SchedulerPointer class
* Adapt code (now System pointer would be ISchedulable pointer)
* Use Template functions to create objects

### CODE REVIEW

File: cocos/core/System.h [hunk 1]
```diff
@@ -32,8 +32,35 @@ namespace cc {
 namespace core {
 
 struct ISchedulable {
+private:
     std::string id;
     std::string uuid;
+
+public:
+    virtual std::string getId() const { return id; }
```
Comment (minggo) [inline cocos/core/System.h:40]: string 类型应该是返回 const std::string &

File: cocos/core/System.h [hunk 2]
```diff
@@ -32,8 +32,35 @@ namespace cc {
...
+    virtual void        setId(const std::string &_id) { id = _id; }
+    virtual std::string getUuid() const { return uuid; }
```
Comment (minggo) [inline cocos/core/System.h:42]: 同上

File: cocos/core/System.h [hunk 3]
```diff
@@ -32,8 +32,35 @@ namespace cc {
...
+    virtual void        setUuid(const std::string &_uuid) { uuid = _uuid; }
+    virtual void        update(float deltaTime) {}
+};
+template <class T>
+class SchedulerPointer : public ISchedulable {
+    T* obj;
+    std::string getId()const override{
+        return T->getId();
+    }
+    void setId(const std:;string &_id) override{
```
Comment (dumganhar) [inline cocos/core/System.h:52]: `:; -> ::`

File: cocos/core/System.h [hunk 4]
```diff
@@ -32,8 +32,35 @@ namespace cc {
...
+    virtual void        setId(const std::string &_id) { id = _id; }
+    virtual std::string getUuid() const { return uuid; }
+    virtual void        setUuid(const std::string &_uuid) { uuid = _uuid; }
+    virtual void        update(float deltaTime) {}
+};
+template <class T>
+class SchedulerPointer : public ISchedulable {
+    T* obj;
+    std::string getId()const override{
+        return T->getId();
```
Comment (dumganhar) [inline cocos/core/System.h:50]: `T->getId()` -> `obj->getId()`?

File: cocos/core/Scheduler.cpp [hunk 5]
```diff
@@ -104,16 +104,16 @@ void TimerTargetCallback::update(float dt) {
...
     Scheduler *            scheduler,
     const ccSchedulerFunc &callback,
-    System *               target,
+    T *                    target,
     float                  interval,
     uint32_t               repeat,
     float                  delay) {
     _lock      = false;
-    _scheduler = scheduler;
+    _scheduler = new SchedulerPointer<T>(target);
```
Comment (dumganhar) [inline cocos/core/Scheduler.cpp:116]: 在哪里 delete 呢？

**Pull request closed without merging**